--- compiz-0.3.2/src/main.c.aiglx-defaults	2006-09-22 11:44:54.000000000 -0400
+++ compiz-0.3.2/src/main.c	2006-11-10 14:32:24.000000000 -0500
@@ -78,8 +78,8 @@
 CompWindow *lastDamagedWindow = 0;
 
 Bool replaceCurrentWm = FALSE;
-Bool indirectRendering = FALSE;
-Bool strictBinding = FALSE;
+Bool indirectRendering = TRUE;
+Bool strictBinding = TRUE;
 
 #ifdef USE_COW
 Bool useCow = FALSE;
@@ -94,8 +94,10 @@
 	    "[--refresh-rate RATE]\n       "
 	    "[--fast-filter] "
 	    "[--indirect-rendering] "
+	    "[--direct-rendering]\n       "
 	    "[--strict-binding] "
-	    "[--test-mode]\n       "
+	    "[--xgl-binding] "
+	    "[--test-mode]\n       " 
 	    "[--replace] "
 	    "[--sm-disable] "
 	    "[--sm-client-id ID] "
