--- compiz-0.3.4/src/main.c.aiglx-defaults	2006-11-11 13:11:36.000000000 -0500
+++ compiz-0.3.4/src/main.c	2006-11-23 01:25:37.000000000 -0500
@@ -78,8 +78,8 @@
 CompWindow *lastDamagedWindow = 0;
 
 Bool replaceCurrentWm = FALSE;
-Bool indirectRendering = FALSE;
-Bool strictBinding = FALSE;
+Bool indirectRendering = TRUE;
+Bool strictBinding = TRUE;
 Bool noDetection = FALSE;
 
 #ifdef USE_COW
@@ -95,10 +95,13 @@
 	    "[--refresh-rate RATE]\n       "
 	    "[--fast-filter] "
 	    "[--indirect-rendering] "
+	    "[--direct-rendering]\n       "
 	    "[--strict-binding] "
-	    "[--replace]\n       "
+	    "[--xgl-binding] "
+	    "[--test-mode]\n       "
+	    "[--replace] "
 	    "[--sm-disable] "
-	    "[--sm-client-id ID] "
+	    "[--sm-client-id ID]\n       "
 	    "[--no-detection] "
 	    "[--version]\n       "
 
