--- compiz-0.3.6/src/main.c~	2006-12-28 21:56:41.000000000 -0500
+++ compiz-0.3.6/src/main.c	2007-01-12 18:06:57.000000000 -0500
@@ -78,7 +78,7 @@
 CompWindow *lastDamagedWindow = 0;
 
 Bool replaceCurrentWm = FALSE;
-Bool indirectRendering = FALSE;
+Bool indirectRendering = TRUE;
 Bool strictBinding = TRUE;
 Bool noDetection = FALSE;
 
@@ -94,7 +94,7 @@
 	    "[--bg-image PNG] "
 	    "[--refresh-rate RATE]\n       "
 	    "[--fast-filter] "
-	    "[--indirect-rendering] "
+	    "[--direct-rendering] "
 	    "[--loose-binding] "
 	    "[--replace]\n       "
 	    "[--sm-disable] "
@@ -188,9 +188,9 @@
 	{
 	    defaultTextureFilter = "Fast";
 	}
-	else if (!strcmp (argv[i], "--indirect-rendering"))
+	else if (!strcmp (argv[i], "--direct-rendering"))
 	{
-	    indirectRendering = TRUE;
+	    indirectRendering = FALSE;
 	}
 	else if (!strcmp (argv[i], "--loose-binding"))
 	{
