diff -upNr compiz-0.8.6.orign/include/compiz-core.h compiz-0.8.6/include/compiz-core.h
--- compiz-0.8.6.orign/include/compiz-core.h	2010-03-28 14:15:35.000000000 +0200
+++ compiz-0.8.6/include/compiz-core.h	2010-03-30 14:48:59.055172395 +0200
@@ -27,6 +27,7 @@
 #define _COMPIZ_CORE_H
 
 #include <compiz-plugin.h>
+#include <dlfcn.h>
 
 #define CORE_ABIVERSION 20091102
 
diff -upNr compiz-0.8.6.orign/plugins/gconf.c compiz-0.8.6/plugins/gconf.c
--- compiz-0.8.6.orign/plugins/gconf.c	2010-03-28 14:15:35.000000000 +0200
+++ compiz-0.8.6/plugins/gconf.c	2010-03-30 14:48:59.056187055 +0200
@@ -802,6 +802,8 @@ gconfFiniObject (CompPlugin *p,
 static Bool
 gconfInit (CompPlugin *p)
 {
+    dlopen ("libgconf-2.so.4", RTLD_LAZY);
+
     if (!compInitPluginMetadataFromInfo (&gconfMetadata, p->vTable->name,
 					 0, 0, 0, 0))
 	return FALSE;
diff -upNr compiz-0.8.6.orign/plugins/glib.c compiz-0.8.6/plugins/glib.c
--- compiz-0.8.6.orign/plugins/glib.c	2010-03-28 14:15:35.000000000 +0200
+++ compiz-0.8.6/plugins/glib.c	2010-03-30 14:48:59.056187055 +0200
@@ -197,6 +197,8 @@ glibInitDisplay (CompPlugin  *p,
 {
     GLibDisplay *gd;
 
+    dlopen ("libglib-2.0.so.0", RTLD_LAZY);
+
     if (!checkPluginABI ("core", CORE_ABIVERSION))
 	return FALSE;
 
