--- compiz-with-glib-mainloop/cmake/plugin_extensions/CompizGenGconf.cmake	2011-01-15 13:20:16.373169983 +0000
+++ compiz-with-glib-mainloop/cmake/plugin_extensions/CompizGenGconf.cmake.new	2011-01-15 16:19:59.265021752 +0000
@@ -63,7 +63,7 @@
 find_program (XSLTPROC_EXECUTABLE xsltproc)
 mark_as_advanced (FORCE XSLTPROC_EXECUTABLE)
 
-if (XSLTPROC_EXECUTABLE AND NOT COMPIZ_DISABLE_SCHEMAS_INSTALL)
+if (XSLTPROC_EXECUTABLE)
     compiz_gconf_prepare_install_dirs ()
     add_custom_command (
 	OUTPUT "${CMAKE_BINARY_DIR}/generated/compiz-${COMPIZ_CURRENT_PLUGIN}.schemas"
