diff -ur compiz-0.7.6/kde/window-decorator-kde4/switcher.cpp compiz-0.7.6-kde42/kde/window-decorator-kde4/switcher.cpp
--- compiz-0.7.6/kde/window-decorator-kde4/switcher.cpp	2008-05-29 13:58:33.000000000 +0200
+++ compiz-0.7.6-kde42/kde/window-decorator-kde4/switcher.cpp	2008-12-10 03:24:48.000000000 +0100
@@ -30,7 +30,7 @@
 
 #include <fixx11h.h>
 
-#include <KDE/Plasma/PanelSvg>
+#include <KDE/Plasma/FrameSvg>
 #include <KDE/Plasma/Theme>
 
 #include <kwindowsystem.h>
@@ -46,9 +46,9 @@
     QColor   color;
     color = Plasma::Theme::defaultTheme ()->color (Plasma::Theme::TextColor);
 
-    mBackground = new Plasma::PanelSvg();
+    mBackground = new Plasma::FrameSvg();
     mBackground->setImagePath ("dialogs/background");
-    mBackground->setEnabledBorders(Plasma::PanelSvg::AllBorders);
+    mBackground->setEnabledBorders(Plasma::FrameSvg::AllBorders);
 
     mBorder.left   = mBackground->marginSize(Plasma::LeftMargin);
     mBorder.right  = mBackground->marginSize(Plasma::RightMargin);
@@ -131,8 +131,8 @@
     p.setCompositionMode (QPainter::CompositionMode_Source);
     p.setRenderHint (QPainter::SmoothPixmapTransform);
 
-    mBackground->resizePanel (QSizeF (contentWidth, contentHeight));
-    mBackground->paintPanel (&p, QRect (0, 0, contentWidth,
+    mBackground->resizeFrame (QSizeF (contentWidth, contentHeight));
+    mBackground->paintFrame (&p, QRect (0, 0, contentWidth,
 			contentHeight));
 
     mBackgroundPixmap = mPixmap.copy (mBorder.left, mBorder.top,
diff -ur compiz-0.7.6/kde/window-decorator-kde4/switcher.h compiz-0.7.6-kde42/kde/window-decorator-kde4/switcher.h
--- compiz-0.7.6/kde/window-decorator-kde4/switcher.h	2008-05-29 13:58:33.000000000 +0200
+++ compiz-0.7.6-kde42/kde/window-decorator-kde4/switcher.h	2008-12-10 03:24:48.000000000 +0100
@@ -30,7 +30,7 @@
 
 namespace Plasma
 {
-class PanelSvg;
+class FrameSvg;
 }
 
 class QSpacerItem;
@@ -70,7 +70,7 @@
 
 	QRect mGeometry;
 
-	Plasma::PanelSvg *mBackground;
+	Plasma::FrameSvg *mBackground;
 	QPixmap mPixmap;
 	QPixmap mBackgroundPixmap;
 
