diff -upNr compiz-0.7.6.orign/plugins/decoration.c compiz-0.7.6/plugins/decoration.c
--- compiz-0.7.6.orign/plugins/decoration.c	2008-05-29 13:58:17.000000000 +0200
+++ compiz-0.7.6/plugins/decoration.c	2008-06-06 12:48:33.000000000 +0200
@@ -831,7 +831,7 @@ decorWindowUpdate (CompWindow *w,
 	moveDy = -oldShiftY;
     }
 
-    if (!w->attrib.override_redirect && (moveDx || moveDy))
+    if (w->placed && !w->attrib.override_redirect && (moveDx || moveDy))
     {
 	XWindowChanges xwc;
 	unsigned int   mask = CWX | CWY;
