diff -upNr compiz-0.8.4.orign/configure.ac compiz-0.8.4/configure.ac
--- compiz-0.8.4.orign/configure.ac	2009-10-14 01:51:02.000000000 +0200
+++ compiz-0.8.4/configure.ac	2010-02-13 21:24:59.644078004 +0100
@@ -517,7 +517,7 @@ if test "x$use_kde4" = "xyes"; then
 			[use_kde4=yes], [use_kde4=no])
 
     KDE4_CFLAGS="-I$kde4incs"
-    KDE4_LIBS="-L$kde4libs -lkdecore -lkdecorations -lplasma"
+    KDE4_LIBS="-L$kde4libs -lkdecore -lkdecorations -lplasma -lX11 -lXext -ldl -lkdeui"
 
     QT4_MOC=`pkg-config QtGui --variable=moc_location`
     QDBUSXML2CPP=`pkg-config QtGui --variable=exec_prefix`"/bin/qdbusxml2cpp"
diff -upNr compiz-0.8.4.orign/src/Makefile.am compiz-0.8.4/src/Makefile.am
--- compiz-0.8.4.orign/src/Makefile.am	2009-09-30 14:46:45.000000000 +0200
+++ compiz-0.8.4/src/Makefile.am	2010-02-13 21:23:26.691078634 +0100
@@ -9,7 +9,7 @@ INCLUDES =			     \
 
 bin_PROGRAMS = compiz
 
-compiz_LDADD = @COMPIZ_LIBS@ @GL_LIBS@ -lm
+compiz_LDADD = @COMPIZ_LIBS@ @GL_LIBS@ -lm -lX11 -lXext -ldl
 compiz_LDFLAGS = -export-dynamic -pthread
 compiz_SOURCES =   \
 	main.c	   \
