--- compiz-with-glib-mainloop/cmake/CompizCommon.cmake	2011-01-23 10:26:25.896687232 -0800
+++ compiz-with-glib-mainloop/cmake/CompizCommon.cmake.new	2011-01-23 17:25:58.739038552 -0800
@@ -42,11 +42,11 @@
 option (COMPIZ_SIGN_WARNINGS "Should compiz use -Wsign-conversion during compilation." OFF)
 
 if (COMPIZ_SIGN_WARNINGS)
-    set (CMAKE_CXX_FLAGS "-Wall -Wsign-conversion")
-    set (CMAKE_C_FLAGS "-Wall -Wsign-conversion")
+    set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wsign-conversion")
+    set (CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wall -Wsign-conversion")
 else ()
-    set (CMAKE_CXX_FLAGS "-Wall")
-    set (CMAKE_C_FLAGS "-Wall")
+    set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall")
+    set (CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wall")
 endif ()
 
 function (compiz_ensure_linkage)
