From f101e9ed8771a3aa32de641b1a3bb43f208812c9 Mon Sep 17 00:00:00 2001
From: Danny Baumann <dannybaumann@web.de>
Date: Fri, 04 Sep 2009 06:10:13 +0000
Subject: Add protection against empty strings (cairo doesn't like 0x0 sized

surfaces).
---
diff --git a/text.c b/text.c
index 723772b..0c1ee7e 100644
--- a/text.c
+++ b/text.c
@@ -284,8 +284,13 @@ textUpdateSurface (CompScreen      *s,
     Display *dpy = s->display->display;
 
     cairo_surface_destroy (data->surface);
+    data->surface = NULL;
+
     cairo_destroy (data->cr);
+    data->cr = NULL;
+
     XFreePixmap (dpy, data->pixmap);
+    data->pixmap = None;
 
     return textInitCairo (s, data, width, height);
 }
@@ -392,7 +397,7 @@ textRenderText (CompScreen           *s,
     TextSurfaceData surface;
     CompTextData    *retval = NULL;
 
-    if (!text)
+    if (!text || !strlen (text))
 	return NULL;
 
     memset (&surface, 0, sizeof (TextSurfaceData));
@@ -448,7 +453,7 @@ textRenderWindowTitle (CompScreen           *s,
 		       const CompTextAttrib *attrib)
 {
     char         *text = NULL;
-    CompTextData *retval;
+    CompTextData *retval = NULL;
 
     if (withViewportNumber)
     {
@@ -480,11 +485,11 @@ textRenderWindowTitle (CompScreen           *s,
 	text = textGetWindowName (s->display, window);
     }
 
-    if (!text)
-	return NULL;
+    if (text && strlen (text))
+	retval = textRenderText (s, text, attrib);
 
-    retval = textRenderText (s, text, attrib);
-    free (text);
+    if (text)
+	free (text);
 
     return retval;
 }
--
cgit v0.8.2

