--- compiz-0.3.6/plugins/cube.c.foo	2007-01-12 18:21:50.000000000 -0500
+++ compiz-0.3.6/plugins/cube.c	2007-01-12 18:22:25.000000000 -0500
@@ -46,14 +46,14 @@
 
 #include <compiz.h>
 
-#define CUBE_COLOR_RED_DEFAULT   0xefef
-#define CUBE_COLOR_GREEN_DEFAULT 0xebeb
-#define CUBE_COLOR_BLUE_DEFAULT  0xe7e7
+#define CUBE_COLOR_RED_DEFAULT   0x0000
+#define CUBE_COLOR_GREEN_DEFAULT 0x1919
+#define CUBE_COLOR_BLUE_DEFAULT  0x4a4a
 
 #define CUBE_IN_DEFAULT FALSE
 
 static char *cubeImages[] = {
-    "freedesktop"
+    "fedora-logo"
 };
 #define N_CUBE_IMAGES (sizeof (cubeImages) / sizeof (cubeImages[0]))
 
--- compiz-0.3.6/plugins/compiz.schemas.in.foo	2006-12-28 22:27:00.000000000 -0500
+++ compiz-0.3.6/plugins/compiz.schemas.in	2007-01-12 18:21:50.000000000 -0500
@@ -15753,7 +15753,7 @@
             <applyto>/apps/compiz/plugins/cube/screen0/options/color</applyto>
             <owner>compiz</owner>
             <type>string</type>
-            <default>#efebe7ff</default>
+            <default>#00194aff</default>
             <locale name="C">
                 <short>Cube Color</short>
                 <long>Color of top and bottom sides of the cube</long>
@@ -15946,7 +15946,7 @@
             <owner>compiz</owner>
             <type>list</type>
             <list_type>string</list_type>
-            <default>[freedesktop]</default>
+            <default>[/usr/share/pixmaps/fedora-logo]</default>
             <locale name="C">
                 <short>Image files</short>
                 <long>List of PNG and SVG files that should be rendered on top face of cube</long>
