diff -upNr compiz-0.8.6.orign/gtk/gnome/Makefile.am compiz-0.8.6/gtk/gnome/Makefile.am
--- compiz-0.8.6.orign/gtk/gnome/Makefile.am	2010-03-28 14:15:35.000000000 +0200
+++ compiz-0.8.6/gtk/gnome/Makefile.am	2010-07-15 20:51:40.241254711 +0200
@@ -1,21 +1,9 @@
 moduledir = $(DESTDIR)$(windowsettingslibdir)/window-manager-settings
 
-if USE_GNOME
-libcompiz_la_LDFLAGS = -export-dynamic -avoid-version -rpath $(moduledir)
-libcompiz_la_LIBADD  = @GNOME_WINDOW_SETTINGS_LIBS@
-libcompiz_la_SOURCES = 		\
-	compiz-window-manager.c	\
-	compiz-window-manager.h
-libcompiz_module     = libcompiz.la
-
 desktopfilesdir	      = $(DESTDIR)$(datadir)/applications
 desktopfiles_in_files = compiz.desktop.in
 desktopfiles_files    = $(desktopfiles_in_files:.desktop.in=.desktop)
 
-wmpropertiesdir	      = $(DESTDIR)$(windowsettingsdatadir)/gnome/wm-properties
-wmproperties_in_files = compiz-wm.desktop.in
-wmproperties_files    = $(wmproperties_in_files:.desktop.in=.desktop)
-
 @INTLTOOL_DESKTOP_RULE@
 
 if USE_GNOME_KEYBINDINGS
@@ -28,24 +16,12 @@ xml_files    = $(xml_in_files:.xml.in=.x
 
 endif
 
-install-exec-local:
-	if mkdir -p $(moduledir) && test -w $(moduledir); then ( \
-		$(LIBTOOL) --mode=install $(INSTALL) $(libcompiz_module) \
-		$(moduledir) \
-		); \
-	fi
-
 install-data-local:
 	if mkdir -p $(desktopfilesdir) && test -w $(desktopfilesdir); then ( \
 		$(INSTALL) -m 644 $(desktopfiles_files) \
 		$(desktopfilesdir) \
 		); \
 	fi
-	if mkdir -p $(wmpropertiesdir) && test -w $(wmpropertiesdir); then ( \
-		$(INSTALL) -m 644 $(wmproperties_files) \
-		$(wmpropertiesdir) \
-		); \
-	fi
 
 if USE_GNOME_KEYBINDINGS
 	if mkdir -p $(xmldir) && test -w $(xmldir); then ( \
@@ -74,23 +50,18 @@ if USE_GNOME_KEYBINDINGS
 		); \
 	fi
 endif
-endif
 
 INCLUDES = @GNOME_WINDOW_SETTINGS_CFLAGS@			   \
 	-DMETACITY_THEME_DIR=\""$(windowsettingsdatadir)/themes"\" \
 	-DLOCALEDIR="\"@datadir@/locale"\"
 
-noinst_LTLIBRARIES = \
-	$(libcompiz_module)
 
 noinst_DATA =		      \
 	$(desktopfiles_files) \
-	$(wmproperties_files) \
 	$(xml_files)
 
 EXTRA_DIST =			 \
 	$(desktopfiles_in_files) \
-	$(wmproperties_in_files) \
 	$(xml_in_files)
 
 DISTCLEANFILES = \
