commit 0ddeef41dd42e22ed75c68ac67154b8de00f9595
Author: Kristian Høgsberg <krh@redhat.com>
Date:   Tue Aug 14 18:28:05 2007 -0400

    Add run_command to keyboard shortcut list.

diff --git a/gtk/gnome/50-compiz-desktop-key.xml.in b/gtk/gnome/50-compiz-desktop-key.xml.in
index da1f8cf..1559d03 100644
--- a/gtk/gnome/50-compiz-desktop-key.xml.in
+++ b/gtk/gnome/50-compiz-desktop-key.xml.in
@@ -11,4 +11,6 @@
 
 	<KeyListEntry name="/apps/compiz/general/allscreens/options/show_desktop_key"/>
 
+        <KeyListEntry name="/apps/compiz/general/allscreens/options/run_command_terminal_key"/>
+
 </KeyListEntries>
