diff -up compiz-6b86f3/metadata/scale.xml.in.scale-key compiz-6b86f3/metadata/scale.xml.in
--- compiz-6b86f3/metadata/scale.xml.in.scale-key	2007-08-22 18:57:43.000000000 -0400
+++ compiz-6b86f3/metadata/scale.xml.in	2007-10-12 10:59:02.000000000 -0400
@@ -12,7 +12,6 @@
 		<allowed key="true" button="true" edge="true" edgednd="true"/>
 		<default>
 		    <key>&lt;Shift&gt;&lt;Alt&gt;Up</key>
-		    <edges top_right="true"/>
 		</default>
 	    </option>
 	    <option name="initiate_all" type="action">
diff -up compiz-6b86f3/gtk/gnome/50-compiz-desktop-key.xml.in.scale-key compiz-6b86f3/gtk/gnome/50-compiz-desktop-key.xml.in
--- compiz-6b86f3/gtk/gnome/50-compiz-desktop-key.xml.in.scale-key	2007-10-12 11:18:03.000000000 -0400
+++ compiz-6b86f3/gtk/gnome/50-compiz-desktop-key.xml.in	2007-10-12 11:16:56.000000000 -0400
@@ -13,4 +13,6 @@
 
         <KeyListEntry name="/apps/compiz/general/allscreens/options/run_command_terminal_key"/>
 
+        <KeyListEntry name="/apps/compiz/plugins/scale/allscreens/options/initiate_key"/>
+
 </KeyListEntries>
