--- compiz-3173a142efe5e7af83ebb534a074d9d2a0c67a86/plugins/move.c.move-terminate	2006-09-28 15:42:09.000000000 -0400
+++ compiz-3173a142efe5e7af83ebb534a074d9d2a0c67a86/plugins/move.c	2006-09-28 15:42:29.000000000 -0400
@@ -365,6 +365,20 @@
     MOVE_DISPLAY (d);
 
     switch (event->type) {
+    case ButtonPress:
+	s = findScreenAtDisplay (d, event->xbutton.root);
+	if (s)
+	{
+	    MOVE_SCREEN (s);
+
+	    if (ms->grabIndex)
+	    {
+		moveTerminate (d,
+			       &md->opt[MOVE_DISPLAY_OPTION_INITIATE].value.action,
+			       0, NULL, 0);
+	    }
+	}
+	break;
     case KeyPress:
     case KeyRelease:
 	s = findScreenAtDisplay (d, event->xkey.root);
