--- compiz-0.4.0/src/screen.c.tfp-server-extension	2007-04-02 05:05:49.000000000 -0400
+++ compiz-0.4.0/src/screen.c	2007-06-05 23:20:52.000000000 -0400
@@ -1396,7 +1396,7 @@ addScreen (CompDisplay *display,
     Window		 *children;
     unsigned int	 nchildren;
     int			 defaultDepth, nvisinfo, nElements, value, i;
-    const char		 *glxExtensions, *glExtensions;
+    const char		 *glxExtensions, *glxServerExtensions, *glExtensions;
     XSetWindowAttributes attrib;
     GLfloat		 globalAmbient[]  = { 0.1f, 0.1f,  0.1f, 0.1f };
     GLfloat		 ambientLight[]   = { 0.0f, 0.0f,  0.0f, 0.0f };
@@ -1645,7 +1645,10 @@ addScreen (CompDisplay *display,
 	    i++;
 
 	glxExtensions = glXQueryExtensionsString (dpy, screenNum);
-	if (!strstr (glxExtensions, "GLX_EXT_texture_from_pixmap"))
+	glxServerExtensions = glXQueryServerString(s->display->display,
+						   screenNum, GLX_EXTENSIONS);
+	if (!strstr (glxExtensions, "GLX_EXT_texture_from_pixmap") &&
+	    !strstr (glxServerExtensions, "GLX_EXT_texture_from_pixmap"))
 	{
 	    if (i > 0)
 	    {
