--- compiz-0.3.6/src/screen.c~	2006-12-29 14:24:40.000000000 -0500
+++ compiz-0.3.6/src/screen.c	2007-01-12 16:01:34.000000000 -0500
@@ -1384,7 +1384,7 @@
     Window		 *children;
     unsigned int	 nchildren;
     int			 defaultDepth, nvisinfo, nElements, value, i;
-    const char		 *glxExtensions, *glExtensions;
+    const char		 *glxExtensions, *glxServerExtensions, *glExtensions;
     GLint	         stencilBits;
     XSetWindowAttributes attrib;
     GLfloat		 globalAmbient[]  = { 0.1f, 0.1f,  0.1f, 0.1f };
@@ -1630,7 +1630,10 @@
 	    i++;
 
 	glxExtensions = glXQueryExtensionsString (dpy, screenNum);
-	if (!strstr (glxExtensions, "GLX_EXT_texture_from_pixmap"))
+	glxServerExtensions = glXQueryServerString(s->display->display,
+						   screenNum, GLX_EXTENSIONS);
+	if (!strstr (glxExtensions, "GLX_EXT_texture_from_pixmap") &&
+	    !strstr (glxServerExtensions, "GLX_EXT_texture_from_pixmap"))
 	{
 	    if (i > 0)
 	    {
